xbmc/filesystem/FileCache.cpp

```diff
@@ -292,20 +306,25 @@ void CFileCache::Process()
       }
     }
 
-    size_t maxWrite = m_pCache->GetMaxWriteSize(m_chunkSize);
+    const int64_t maxWrite = m_pCache->GetMaxWriteSize(m_chunkSize);
+    int64_t maxSourceRead = m_chunkSize;
+    // Cap source read size by space available between current write position and EOF
+    if (m_fileSize != 0)
+      maxSourceRead = std::min(maxSourceRead, m_fileSize - m_writePos);
 
     /* Only read from source if there's enough write space in the cache
      * else we may keep disposing data and seeking back on (slow) source
      */
-    if (maxWrite < m_chunkSize && !cacheReachEOF)
+    if (maxWrite < maxSourceRead && !cacheReachEOF)
     {
+      // Wait until sufficient cache write space is available
       m_pCache->m_space.WaitMSec(5);
       continue;
     }
 
     ssize_t iRead = 0;
     if (!cacheReachEOF)
```

Comment:
@MilhouseVH : Could you try changing this line into " if (!cacheReachEOF && maxSourceRead > 0) ?

Reply:
I added this patch to Ubuntu:
```
diff --git a/xbmc/filesystem/FileCache.cpp b/xbmc/filesystem/FileCache.cpp
index 1f87385..ab2a686 100644
--- a/xbmc/filesystem/FileCache.cpp
+++ b/xbmc/filesystem/FileCache.cpp
@@ -318,7 +318,8 @@ void CFileCache::Process()
     if (maxWrite < maxSourceRead && !cacheReachEOF)
     {
       // Wait until sufficient cache write space is available
-      m_pCache->m_space.WaitMSec(5);
+      if (!cacheReachEOF && maxSourceRead > 0)
+        m_pCache->m_space.WaitMSec(5);
       continue;
     }

```
but it had no effect - still hangs with buffermode 1 and 4.

Reply:
@MilhouseVH : Sorry, should have been more accurate. You should try this:

diff --git a/xbmc/filesystem/FileCache.cpp b/xbmc/filesystem/FileCache.cpp
index 1f87385685..e35a2c80c3 100644
--- a/xbmc/filesystem/FileCache.cpp
+++ b/xbmc/filesystem/FileCache.cpp
@@ -323,7 +323,7 @@ void CFileCache::Process()
     }
 
     ssize_t iRead = 0;
-    if (!cacheReachEOF)
+    if (!cacheReachEOF && maxSourceRead > 0)
       iRead = m_source.Read(buffer.get(), maxSourceRead);
     if (iRead == 0)
     {

Reply:
Sorry - for some reason I thought you were referring to the green line! Not enough coffee... :)

Success! The change above is working fine with all buffermodes (0/1/2/3/4) and `nfs://` (and also `smb://`) - thanks!

Reply:
@MilhouseVH : Great. Btw. the hanging you experienced shouldn't happen, this also also something that needs attention. Did you also try stopping player during the hang?

Reply:
Yes, kodi is _completely_ hung, and the GUI is unresponsive after seeking (without the above fix) - I had to `kill -9` the kodi process (or reboot).

Reply:
Strange, will look into that. Apperently this happens when a read returns a failure, but the intended design was that it should empty buffer and terminate. Will look into that as well before PR this. @MilhouseVH : would it be too much trouble to get a call-trace for this?

Reply:
Tell me how and I'll get it to you.

Reply:
Is this any good to you? I killed kodi with `kill -SIGSEGV $(pidof kodi.bin)` after it had hung.

http://ix.io/27xG

Reply:
Yes, thanks! I was looking in the wrong direction, since it's not hanging in the FileCache thread but in CDVDInputStreamFile::Seek/XFILE::CFileCache::Seek and I suspect what's causing it but I need to setup an NFS server myself to test it with.

Reply:
@MilhouseVH : Unfortunately I can't reproduce the hang-issue here. Could you please test this patch (without the other one):
```
diff --git a/xbmc/filesystem/FileCache.cpp b/xbmc/filesystem/FileCache.cpp
index 1f87385685..b07542ee6f 100644
--- a/xbmc/filesystem/FileCache.cpp
+++ b/xbmc/filesystem/FileCache.cpp
@@ -363,6 +363,8 @@ void CFileCache::Process()
     {
       CLog::Log(LOGDEBUG, "CFileCache::Process - Source read returned a fatal error! Will wait for buffer to empty.");
 
+      m_pCache->EndOfInput();
+
       while (m_pCache->WaitForData(0, 0) > 0)
       {
         if (m_seekEvent.WaitMSec(100))
```

And let me know whether it still hangs?

Reply:
Re-testing the original issue (without any patches) I appear to have misremembered how the GUI behaves - seeking causes video playback to stall/freeze, but the GUI remains functional (I can bring up the informational display, debug output continues to update etc.). Once I stop the video, only THEN does kodi hang completely (GUI is now frozen).

> Unfortunately I can't reproduce the hang-issue here.

That's a bummer. :(

> Could you please test this patch (without the other one):

With the new patch video playback will stall as soon as I seek, the GUI will remain functional, and then Kodi will freeze once I stop playback so it's the same as without the patch.

Here's an Ubuntu trace (with latest patch) while video playback is frozen but the GUI is still operational: http://ix.io/27zT

And now the same again, but this after pressing stop to end playback (the GUI is now frozen): http://ix.io/27zU

Hope this helps!

Reply:
@MilhouseVH : Just try to reproduce the cache+NFS freeze-issue again with my Mint 19 desktop + Netgear NAS  and I'm finally able to reproduce the issue with one of my ISO-files. It seems there is a race between the FileCache thread terminating and checking of the m_seekEnded-event in CFIleCache::Seek()